feat(eslint-plugin): add `no-duplicate-type-constituents` rule (#5728)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/no-duplicate-type-constituents.md

---
description: 'Disallow duplicate union/intersection type members.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.

> Although TypeScript supports duplicate union and intersection member values, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.

Reviewer: [Docs] Here and elsewhere in the file - we should be consistent with how we phrase it.

```suggestion
Although TypeScript supports duplicate union and intersection constituents, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.
```

## Rule Details

> This rule disallows duplicate union or intersection type members. It only checks duplication on the notation. Members with the same value but different names are not marked duplicates.

Reviewer: [Docs] I don't think this is accurate anymore?

I also don't know that people would know what _"same value"_ means. Could you elaborate?
* We do check whether type equality ends up being the same
* But sometimes TypeScript treats things that are the same type as different ones:
    ```ts
    type WrapInValue<T> = { value: T };
    type T = { value: number } | WrapInValue<number>;
    ``` 

I'd suggest adding to the docs some explanation of how this rule determines whether two types are equivalent: that we check whether the syntax is the same, or if TypeScript treats them as the exact same type. This would both help explain edge cases to users and help explain why the code is doing _two_ equality checks.

Yet again I am wishing for https://github.com/microsoft/TypeScript/issues/9879 🥲 which would allow us to skip all this logic, and maybe even combine the rule with `no-redundant-type-constituents`...

---
description: 'Disallow duplicate constituents of union or intersection types.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-duplicate-type-constituents** for documentation.

> Although TypeScript supports duplicate union and intersection constituents, people usually expect members to have unique values within the same intersection and union. Duplicate values make the code redundant and generally reduce readability.

Reviewer: Just proofreading a bit to make it more clear what a "constituent" is for someone who hasn't heard the term before, like how https://typescript-eslint.io/rules/no-redundant-type-constituents does. How does this make you feel?

```suggestion
Although TypeScript supports types ("constituents") in union and intersection types being duplicates of each other, developers typically expect each constituent to be unique within the same intersection and union.
Duplicate values make the code overly verbose and generally reduce readability.
```

Author: Looks great! 
In fact, I was not familiar with the word `constituents` until I started working on this PR, but I think this sentence is much easier to understand.

Reviewer: Haha yeah I learned a ton of random new words working on linting tools. It's why I recommend them to people as a great type of open source project to dive into!

This PR is shaping up nicely btw. I'm really excited to see it in. Awesome work @sajikix! 👏

## Rule Details

This rule disallows duplicate union or intersection constituents.

> 
> It determines whether two types are equivalent in the following way.
> 
> 1. whether the syntax is exactly the same.
> 2. whether TypeScript treats them as the same type.
> 
> If either of the two conditions is satisfied, It treats the two types as duplicates.

Reviewer: I don't think most users really care about how we determine the equivalency internally. The syntax being exactly the same is an internal optimization to avoid using the type checker if we don't have to.

```suggestion
We consider types to be duplicate if they evaluate to the same result in the type system.
For example, given `type A = string` and `type T = string | A`, this rule would flag `string | A`.
```

Author: Certainly, there was no need to even talk about internal optimization here...

<!--tabs-->

### ❌ Incorrect

```ts
type T1 = 'A' | 'A';

type T2 = A | A | B;

type T3 = { a: string } & { a: string };

type T4 = [1, 2, 3] & [1, 2, 3];

type StringA = string;
type StringB = string;
type T5 = StringA | StringB;
```

### ✅ Correct

```ts
type T1 = 'A' | 'B';

type T2 = A | B | C;

type T3 = { a: string } & { b: string };

> type T4 = [1, 2, 3] & [1, 2, 3, 4];

Reviewer: This evaluates to `never`, so I'd suggest a `union` type instead:

```suggestion
type T4 = [1, 2, 3] | [1, 2, 3, 4];
```

(and probably the _Incorrect_ code case should be a `|` union too)

Author: ✅ done 🙏

type StringA = string;
type NumberB = number;
type T5 = StringA | NumberB;
```

## Options

### `ignoreIntersections`

> When set to true, duplicate checks on intersection type members are ignored.

Reviewer: ```suggestion
When set to true, duplicate checks on intersection type constituents are ignored.
```

Author: ✅ done 🙏

### `ignoreUnions`

> When set to true, duplicate checks on union type members are ignored.

Reviewer: ```suggestion
When set to true, duplicate checks on union type constituents are ignored.
```

Author: ✅ done 🙏
